tensorflow/core/transforms/remapper/remapping_helper.h

```diff
@@ -0,0 +1,229 @@
+limitations under the License.
+==============================================================================*/
+#ifndef TENSORFLOW_CORE_TRANSFORMS_REMAPPER_REMAPPING_HELPER_H_
+#define TENSORFLOW_CORE_TRANSFORMS_REMAPPER_REMAPPING_HELPER_H_
+
+#include "tensorflow/core/transforms/utils/op_cat_helper.h"
+#include "tensorflow/core/transforms/utils/utils.h"
+
+namespace mlir {
+namespace tfg {
+namespace remapping {
+
+// Helper structure for collecting fusion info
+struct ContractionWithBiasAdd {
+  Operation* contraction;
+  Operation* bias_add;
+};
+
+// Helper structure for collecting fusion info
+struct ContractionWithBiasAddAndActivation {
+  Operation* contraction;
+  Operation* bias_add;
+  Operation* activation;
+};
+
+enum class Activation { Relu, Relu6, Elu, LeakyRelu, Tanh, Sigmoid };
+
+inline std::string GetTFGActivation(Activation activation) {
+  switch (activation) {
+    case Activation::Relu:
+      return "tfg.Relu";
+    case Activation::Relu6:
+      return "tfg.Relu6";
+    case Activation::Elu:
+      return "tfg.Elu";
+    case Activation::LeakyRelu:
+      return "tfg.LeakyRelu";
+    case Activation::Tanh:
+      return "tfg.Tanh";
+    case Activation::Sigmoid:
+      return "tfg.Sigmoid";
+    default:
+      return "tfg.NoOp";
+  }
+}
+
+class OpPropertyHelper : public OpCatHelper {
+ public:
+  OpPropertyHelper(TFGraphDialect* dialect, bool onednn_enabled = false,
+                   bool xla_auto_clustering = false)
+      : OpCatHelper(dialect),
+        dialect_(dialect),
+        is_onednn_enabled_(onednn_enabled),
+        xla_auto_clustering_(xla_auto_clustering) {}
+
+  // Get the TFG dialect instance.
+  TFGraphDialect* getDialect() { return dialect_; }
+
+  bool HasControlFaninOrFanOut(Operation* op) {
+    TFOp wrapper_op(op);
+    bool has_ctl_fanin = !(wrapper_op.getControlOperands().empty());
+    bool has_ctl_fanout = !(wrapper_op.controlRet().getUses().empty());
+    if (has_ctl_fanin || has_ctl_fanout)
+      return true;
+    else
+      return false;
+  }
+
+  bool HasAtMostOneFanoutAtPort0(Operation* op) {
+    // Note tfg operations have at least 2 results: at least 1 non-control
+    // and exactly 1 control result.
+    return op->getNumResults() > 0 &&
+           (op->getResult(0).hasOneUse() || op->getResult(0).use_empty());
+  }
+
+  bool IsContraction(Operation* op) {
+    return dialect_->IsConv2D(op) || dialect_->IsConv3D(op) ||
+           dialect_->IsDepthwiseConv2dNative(op) || dialect_->IsMatMul(op) ||
+           dialect_->IsMatMul(op);
+  }
+
+  bool HaveSameDataType(Operation* lhs_op, Operation* rhs_op,
+                        const StringRef& attr_name = "T") {
+    if (!lhs_op->hasAttrOfType<TypeAttr>(attr_name) ||
+        !rhs_op->hasAttrOfType<TypeAttr>(attr_name))
+      return false;
+    Type lhs_dtype = lhs_op->getAttrOfType<TypeAttr>(attr_name).getValue();
+    Type rhs_dtype = rhs_op->getAttrOfType<TypeAttr>(attr_name).getValue();
```

Comment:
You may consider,
auto lhs_attr = lhs_op->getAttrOfType<TypeAttr>(...);
auto rhs_attr = rhs_op->getAttrOfType<TypeAttr>(...);

if (!lhs_attr || !rhs_attr) return false;

Type lhs_dtype = lhs_attr.getValue();
Type rhs_dtype = rhs_attr.getValue();